pymove/tests/core/test_data_frame.py

```diff
@@ -0,0 +1,1102 @@
+    )
+    assert_frame_equal(move_df.describe(), expected)
+
+
+def test_memory_usage():
+    move_df = _default_move_df()
+
+    expected = Series(
+        data=[128, 32, 32, 32, 32],
+        index=['Index', 'lat', 'lon', 'datetime', 'id'], dtype='int64', name=None
+    )
+    assert_series_equal(move_df.memory_usage(), expected)
+
+
+def test_copy():
+    move_df = _default_move_df()
+
+    cp = move_df.copy()
+    assert_frame_equal(move_df, cp)
+    cp.at[0, TRAJ_ID] = 0
+    assert move_df.loc[0, TRAJ_ID] == 1
+    assert move_df.loc[0, TRAJ_ID] != cp.loc[0, TRAJ_ID]
+
+
+def test_generate_tid_based_on_id_datetime():
+    move_df = _default_move_df()
+
+    new_move_df = move_df.generate_tid_based_on_id_datetime(inplace=False)
+    expected = DataFrame(data=[
+        [39.984094, 116.319236, Timestamp('2008-10-23 05:53:05'), 1, '12008102305'],
+        [39.984198, 116.319322, Timestamp('2008-10-23 05:53:06'), 1, '12008102305'],
+        [39.984224, 116.319402, Timestamp('2008-10-23 05:53:11'), 2, '22008102305'],
+        [39.984224, 116.319402, Timestamp('2008-10-23 05:53:11'), 2, '22008102305'],
+    ], columns=['lat', 'lon', 'datetime', 'id', 'tid'], index=[0, 1, 2, 3]
+    )
+    assert_frame_equal(new_move_df, expected)
+    assert isinstance(new_move_df, PandasMoveDataFrame)
+    assert TID not in move_df
+
+    move_df.generate_tid_based_on_id_datetime()
+    assert_frame_equal(move_df, expected)
+
+
+def test_generate_date_features():
+    move_df = _default_move_df()
+
+    new_move_df = move_df.generate_date_features(inplace=False)
+    expected = DataFrame(data=[
+        [39.984094, 116.319236, Timestamp('2008-10-23 05:53:05'), 1, date(2008, 10, 23)],
+        [39.984198, 116.319322, Timestamp('2008-10-23 05:53:06'), 1, date(2008, 10, 23)],
+        [39.984224, 116.319402, Timestamp('2008-10-23 05:53:11'), 2, date(2008, 10, 23)],
+        [39.984224, 116.319402, Timestamp('2008-10-23 05:53:11'), 2, date(2008, 10, 23)],
+    ], columns=['lat', 'lon', 'datetime', 'id', 'date'], index=[0, 1, 2, 3]
+    )
+    assert_frame_equal(new_move_df, expected)
+    assert isinstance(new_move_df, PandasMoveDataFrame)
+    assert DATE not in move_df
+
+    move_df.generate_date_features()
+    assert_frame_equal(move_df, expected)
+
+
+def test_generate_hour_features():
+    move_df = _default_move_df()
+
+    new_move_df = move_df.generate_hour_features(inplace=False)
+    expected = DataFrame(data=[
+        [39.984094, 116.319236, Timestamp('2008-10-23 05:53:05'), 1, 5],
+        [39.984198, 116.319322, Timestamp('2008-10-23 05:53:06'), 1, 5],
+        [39.984224, 116.319402, Timestamp('2008-10-23 05:53:11'), 2, 5],
+        [39.984224, 116.319402, Timestamp('2008-10-23 05:53:11'), 2, 5],
+    ], columns=['lat', 'lon', 'datetime', 'id', 'hour'], index=[0, 1, 2, 3]
+    )
+    assert_frame_equal(new_move_df, expected)
+    assert isinstance(new_move_df, PandasMoveDataFrame)
+    assert HOUR not in move_df
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5574862)